cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,678 @@
+	}
+
+	return nil
+}
+
+func makeTar(cmd *cobra.Command, t *TarPipe) error {
+
+	// making Tar flow:
+	// tar --> gzip --> buf
+
+	compress, err := cmd.Flags().GetBool("compress")
+	if err != nil {
+		return err
+	}
+
+	var writer io.Writer
+	writer = t.outStream
+
+	if compress {
+		zr := gzip.NewWriter(writer)
+		defer zr.Close()
+		writer = zr
+	}
+
+	tw := tar.NewWriter(writer)
+	defer tw.Close()
+
+	srcDir, srcBase := copyutil.SplitPathDirEntry(filepath.Clean(t.src))
+	dstDir, dstBase := copyutil.SplitPathDirEntry(filepath.Clean(t.dst))
+
+	// At this point, we consider that destination exists and it is a directory
+	// Override dstBase by copying specified src into it
+	dstBase = filepath.Join(dstBase, srcBase)
+
+	return recursiveTar(srcDir, srcBase, dstDir, dstBase, tw)
+}
+
+// recursiveTar makes a tar recursively
+func recursiveTar(srcDir, srcBase, dstDir, dstBase string, tw *tar.Writer) error {
+	matchedPaths, err := filepath.Glob(filepath.Join(srcDir, srcBase))
+	if err != nil {
+		return err
+	}
+
+	for _, path := range matchedPaths {
+		stat, err := os.Lstat(path)
+		if err != nil {
+			return err
+		}
+		if stat.IsDir() {
+			files, err := os.ReadDir(path)
+			if err != nil {
+				return err
+			}
+			if len(files) == 0 {
+				//case empty directory
+				hdr, _ := tar.FileInfoHeader(stat, path)
+				hdr.Name = dstBase
+				if err := tw.WriteHeader(hdr); err != nil {
+					return err
+				}
+			}
+
+			for _, f := range files {
+				if err := recursiveTar(srcDir, filepath.Join(srcBase, copyutil.StripTrailingSlash(f.Name())),
+					dstDir, filepath.Join(dstBase, copyutil.StripTrailingSlash(f.Name())), tw); err != nil {
+					return err
+				}
+			}
+			return nil
+		} else if stat.Mode()&os.ModeSymlink != 0 {
+			//case soft link
+			hdr, _ := tar.FileInfoHeader(stat, path)
+			target, err := os.Readlink(path)
+			if err != nil {
+				return err
+			}
+
+			hdr.Linkname = target
+			hdr.Name = dstBase
+			if err := tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+		} else {
+			//case regular file or other file type like pipe
+			hdr, err := tar.FileInfoHeader(stat, path)
+			if err != nil {
+				return err
+			}
+			hdr.Name = dstBase
+
+			if err := tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+
+			f, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+
+			if _, err := io.Copy(tw, f); err != nil {
+				return err
+			}
+
+			return f.Close()
+		}
+		return nil
+	}
+
+	return nil
+}
+func copyFromContainer(ctx context.Context, cmd *cobra.Command, copyConfig cpConfig) (err error) {
+	dstPath := copyConfig.destPath
+	srcPath := copyConfig.sourcePath
+
+	retries, err := cmd.Flags().GetInt("retries")
+	if err != nil {
+		return err
+	}
+
+	compress, err := cmd.Flags().GetBool("compress")
+	if err != nil {
+		return err
+	}
+
+	t := initTarPipe(retries, &copyConfig)
+
+	// check if tar binary exists in the container
+	testDCmd := []string{"which", "tar"}
```

Comment:
`which` is often missing.

And I guess we can just use the host `tar` or just `cp`

Reply:
>  I don't see how we can use the host tar binary

With /proc/PID/root .